### PR TITLE
[MRG] Successive halving

### PR DESCRIPTION
**Pull request by NicolasHug**
~~Still very WIP, need feedback on the interface :)~~

I'm now overwriting `_run_search`, taking advantage of the proposed changes in `BaseGridSearch` from https://github.com/scikit-learn/scikit-learn/pull/13145

Note that this forces to remove the `refit` parameter, which has to be set to a custom callable.

`cv_results_` looks good and it's still useful. For e.g. 4 candidates it's going to have 4 (first iter) + 2 (second and last iter) rows.

If you're OK with this new design I'll start to write tests :)

### CODE REVIEW

**Comment by amueller** [comment]
missing
```python
from collections import defaultdict
from itertools import product
```
I think?

**Comment by amueller** [comment]
in the random search it probably shouldn't be n_iter but n_candidates?

Also there should be some minimum size for the data for the first iteration.

**Comment by amueller** [comment]
I wonder if we should be using a bigger validation set in the cross-validation... hm... I guess large validation sets could also slow things down. Have you checked how other implementations do this?

**Comment by amueller** [comment]
```python
res = pd.DataFrame(sh.cv_results_)
res['params_str'] = res.params.apply(str)
reshape = res.pivot(index='iter', columns='params_str', values='mean_test_score')
reshape.plot(legend=False, alpha=.4, c='k')
```
image

File: fml/search.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        candidate_params = list(self._generate_candidate_params())
+        n_iterations = int(ceil(log2(len(candidate_params))))
+        n_samples_total = X.shape[0]
+
+        for iter_i in range(n_iterations):
+            # randomly sample training samples
+            n_candidates = len(candidate_params)
+            n_samples_iter = floor(n_samples_total /
+                                   (n_candidates * n_iterations))
+            indices = rng.choice(n_samples_total, n_samples_iter,
```
Comment (amueller) [inline fml/search.py]:
```suggestion
			# this could be outside the for-loop
			# 2 is a magic number. I found 10 too slow and 2 seems to work fine?
			# basically lower bound on the test set size
			cv = check_cv(self.cv, y, classifier=is_classifier(self.estimator))
			min_n_samples = cv.get_n_splits(X, y) * n_classes * 2

            if is_classifier(self.estimator):
                n_samples_iter = max(n_samples_iter, min_n_samples)
            indices = rng.choice(n_samples_total, n_samples_iter,
```

this makes this more robust for small datasets.

Comment (amueller) [reply fml/search.py]: ok the is_classifier here is useless and my indentation is wrong... don't commit this lol.

File: fml/search.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                         return_train_score=return_train_score, iid=False)
+
+        self.random_state = random_state
+
+    def _run_search(self, evaluate_candidates, X, y):
+        rng = check_random_state(self.random_state)
+
+        candidate_params = list(self._generate_candidate_params())
+        n_iterations = int(ceil(log2(len(candidate_params))))
+        n_samples_total = X.shape[0]
```
Comment (amueller) [inline fml/search.py]: ok so you basically set the budget to n_samples, right? where did you get that from?

**Comment by amueller** [comment]
tests are passing now on master :)

**Comment by amueller** [comment]
btw if you can make this green I'll merge it and we can iterate. That way I can play with it on my flight more easily.

**Comment by NicolasHug** [comment]
CI is queued but it should be green now (obligatory *it works on my machine* ;))

**Comment by amueller** [comment]
thanks!
